pkg/sfu/downtrack.go

```diff
@@ -565,6 +579,34 @@ func (d *DownTrack) keyFrameRequester(generation uint32, layer int32) {
 	}
 }
 
+func (d *DownTrack) postMaxLayerNotifierEvent(force bool) {
+	if !force && d.IsClosed() {
+		return
+	}
+
+	select {
+	case d.maxLayerNotifierCh <- struct{}{}:
+	default:
+		d.logger.Warnw("max layer notifier event queue full", nil)
+	}
+}
+
+func (d *DownTrack) maxLayerNotifierWorker() {
+	more := true
+	for more {
+		_, more = <-d.maxLayerNotifierCh
+
+		maxLayerSpatial := buffer.InvalidLayerSpatial
+		if more {
+			maxLayerSpatial = d.forwarder.GetMaxSubscribedSpatial()
+		}
+		if onMaxSubscribedLayerChanged := d.getOnMaxLayerChanged(); onMaxSubscribedLayerChanged != nil {
+			d.logger.Infow("max subscribed layer changed", maxLayerSpatial)
+			onMaxSubscribedLayerChanged(d, maxLayerSpatial)
+		}
+	}
+}
+
```

Comment:
This is the main change for the out-of-order max layer notification which was causing some dynacast test failures. Basically, in places where max subscribed layer changes (mute/unmute, explicit subscribed layer change, when layer switches down, down track close), an event is posted and current state is assessed to figure out the max subscribed layer to report.